### PR TITLE
Allow field value functions to be null/invalid

### PR DESCRIPTION
**Pull request by jamesattewell**
Hi @m-peko

This pull request allows field value functions passed into evaluator to throw booleval::field_null exception which evaluates to false in result_vistor::visit_logical().

I also added some tests for the new exception mechanism covering both exceptions.

The booleval::field_null exception also needs including in the docs and an example which I will do once I know you are happy with the implementation.

Cheers

James

### CODE REVIEW

**Comment by m-peko** [comment]
Can you please explain me use case for this? Where can this be applied?

**Comment by jamesattewell** [comment]
It's the same as allowing nulls in a database when there isn't a safe default value.

In my use case, I have data flowing in from a CANbus and some data items will take time to have valid data and some will never become valid.  There aren't any safe defaults I can use for these values. I need to be able to evaluate the expressions and have them fail if the data isn't valid.

I can't think of another simple way to handle this sort of use case without allowing fields to have a null or invalid value. Is there another way you can think of to do this?

Cheers

James

On 5 Oct 2020 20:30, Marin Peko <notifications@github.com> wrote:

Can you please explain me use case for this? Where can this be applied?

**Comment by m-peko** [comment]
So can you write down how your expression would look like and how do the objects sent to evaluator look like? Example could be a simple `Foo`, `Bar`... thing :)

**Comment by jamesattewell** [comment]
I have a CAN database class which is connected to the CANbus and this decodes data in the frames to member variables (int, uint, string etc, etc). Each of these has a valid flag to indicate if they have been set from the CAN, this valid flag can also be reset after a timeout if the data hasn't been refreshed. I need to be able to evaluate expressions when these CAN database items aren't valid and the expression needs to fail if the data/field isn't valid. Unfortunately, some of the items from CAN don't have any safe default values I can use and return from the field_map functions.

So the expressions are simple i.e. "(canVar1 == 10) && (canVar2 > 5)"

I don't like the exception mechanism, its clumsy and not very efficient, but it fits without changing the signatures of the evaluator field_map functions. If we could change these to accept a bool reference as a parameter to denote if the field is valid this would also work. So rather than:
`uint32_t canVar1() const noexcept`
we have something like this:
`uint32_t canVar1(bool& is_valid) const noexcept`
Where is_valid would be checked by booleval and, if false, would fail the evaluation for the field/variable.

Maybe there is a way this can be achieved using a template mechanism?

I also see the CI is failing with GCC so the code needs fixing anyway. I am developing using VC++ at the moment and it all works with the MS compiler.

Cheers

James

**Comment by m-peko** [comment]
Why don't you set the expression like ` ((canVar1 == 10) && (canVar2 > 5)) or canVar1 == 0`? If the field is not set by the CAN, then it will be zero.

**Comment by jamesattewell** [comment]
Unfortunately, there are some fields from the CANbus that I can't have a default invalid value for such as zero, all possible values are valid.

I think I have another solution to the problem which avoids the field_null exception. I am working on the test harness at the moment and will modify the PR and push shortly. It allows the field_map any_mem_fn to have either no parameter or a bool&.

**Comment by m-peko** [comment]
I don't like this solution because it's not a generic one. It can be applied to your problem but I doubt other library users would benefit from this... You can maybe think of a more generic solution, maybe passing variadic parameter pack to `any_mem_fn` so that it's more generic and not fixed to `is_valid` parameter

**Comment by jamesattewell** [comment]
I agree its not ideal, especially having to have fn_ and fn1_ member variable. I will do some more thinking on it, I was trying to figure out if/how we could pass the any_mem_fn class as a template parameter.

Can you give a simple example of your suggestion with the variadic parameter pack please? I am not quite sure how that would be implemented.

**Comment by m-peko** [comment]
Hi @jamesattewell, sorry for late response. I was working on other projects. So you say all possible values are valid for some fields... You can try the following hack:

Let's say you have the following class representing CAN bus fields:
```c++
template <typename T>
class obj {
public:
    obj() : is_value_a_valid_{false}, value_a_{} {}
    obj(T value) : is_value_a_valid_{true}, value_a_{ value } {}
    
    bool is_value_a_valid const noexcept { return is_value_a_valid_; }
    
    T value_a() const noexcept {
        // ...
        is_value_a_valid_ = false;
        // ...
        return value_a_;
    }

private:
    bool is_value_a_valid_;
    T value_a_;
};
```

and then you can do something like:

```c++
obj<double> baz{ 2.345678 };

booleval::evaluator evaluator({
    { "field_a", &obj<double>::value_a }
});

evaluator.expression("field_a leq 1.234567"));
if (evaluator.is_activated()) {
     if (foo.is_value_a_valid() && evaluator.evaluate(foo)) {
         // do something
     }
}
```

What you think of this?

**Comment by jamesattewell** [comment]
Hi @m-peko, no worries, I know exactly what's it like ;)

Unfortunately, this won't work for me as the expressions are editable outside the application I am using this for. So I would need to pre-parse the expression to call all the is_valid() functions for the fields used in each expression and this sort of defeats the object.

I have been trying to figure out if there is a way we can have evaluator take the any_mem_fn type class as a template parameter, which can be passed down the tree? This would allow other implementations to be used instead, like the version I need with the bool& is_valid parameter. I haven't been able to spend much time on it yet but I am sure there must be a way to do this, any ideas?

**Comment by jamesattewell** [comment]
Hi @m-peko 

I have changed around the code as I figured out a way to do what I wanted with templates. There is a new any_mem_fn_bool class which can be specified as a template parameter to evaluator (it defaults to the original any_mem_fn class).

I am working on adding the quote_char parameter to the evaluator as well (defaulted to double_quote).

File: include/booleval/evaluator.hpp [hunk 1]
```diff
@@ -1,103 +1,127 @@
...
+    [[nodiscard]] bool evaluate(T const& obj) {
+        if (is_activated_) {
+            return result_visitor_.visit(*expression_tree_.root(), obj);
+        } else {
+            return false;
+        }
+    }
+
+private:
+    bool is_activated_;
```
Comment (m-peko) [inline include/booleval/evaluator.hpp:105]: You can make use of in-class initialization here and do: `bool is_activated_{ false };` and then empty ctor would be `evaluator() = default;`

File: src/token/tokenizer.cpp [hunk 2]
```diff
@@ -1,99 +1,103 @@
...
+#include <booleval/token/tokenizer.hpp>
+#include <booleval/utils/string_utils.hpp>
+#include <booleval/utils/split_range.hpp>
+
+namespace booleval {
+
+namespace token {
+
+tokenizer::tokenizer()
+    : current_token_index_(0) {
```
Comment (m-peko) [inline src/token/tokenizer.cpp:39]: Initialize in-class and make ctor `= default`;

File: include/booleval/evaluator.hpp [hunk 3]
```diff
@@ -1,103 +1,127 @@
...
+
+namespace booleval {
+
+/**
+ * class evaluator
+ *
+ * Represents a class for evaluating logical expressions in a form of a string.
+ * It builds an expression tree and traverses that tree in order to evaluate fields.
+ */
+template <typename AnyMemFn = utils::any_mem_fn>
```
Comment (m-peko) [inline include/booleval/evaluator.hpp:47]: I don't like template parameter names to be too long... so maybe `F` would be enough. Please, change this in all parts of code.

**Comment by m-peko** [comment]
I think the solution is good so I will merge this once you fix the comments. Thanks for contributing.

**Comment by jamesattewell** [comment]
Hi @m-peko.

I have implemented your comments above (in class initializer and default constructor) and also fixed up some other classes to do the same. 

New evaluator template parameter name changed to AMF as F caused problems in result_vistor where F is already used by visit_logical() and visit_releational().

I also added a fields() function in evaluator to allow use of its default constructor and setting the fields later on.

Cheers

James

File: include/booleval/tree/tree_node.hpp [hunk 4]
```diff
@@ -45,29 +45,21 @@ namespace tree {
  * as well as the token that the node represents in the actual expression tree.
  */
 struct tree_node {
-    token::token token;
+    token::token token{ token::token_type::unknown };
     std::shared_ptr<tree_node> left;
     std::shared_ptr<tree_node> right;
```
Comment (m-peko) [inline include/booleval/tree/tree_node.hpp:50]: Missing initialization of smart pointer to `nullptr`

**Comment by m-peko** [comment]
Hi @jamesattewell, then maybe I would rather prefer `MemFn` or `Fn` (you pick).

File: include/booleval/tree/tree_node.hpp (hunk 4 above)
Comment (jamesattewell) [reply include/booleval/tree/tree_node.hpp:50]: @m-peko I was under the impression that a shared_ptr will already be nullptr by default and doesn't need initialising?

**Comment by jamesattewell** [comment]
@m-peko I have renamed from AMF to MemFn as suggested. I don't think we need to init the std::shared_ptr to nullptr.

**Pull request merged**